puppet/lib/puppet/util/puppetdb/http.rb

```diff
@@ -222,6 +222,29 @@ def self.reset_query_failover()
       @@last_good_query_server_url_index.reset(0)
     end
 
+    # This function is a bit of a hack to get around the fact that prior to
+    # Puppet Server 5, Puppet Server's http client's `get` method only took
+    # two arguments (both required), whereas the Ruby implementation takes
+    # three arguments (two are optional).
+    #
+    # Since we don't have any guarantees over what version of Puppet/Puppet
+    # Server this terminus code will run under, this method inspects the `get`
+    # method to check its arity, and then calls it with the correct number of
+    # arguments.
+    def self.multi_arity_get(http_instance, path, headers, options={})
+      # `parameters()` returns the parameters of a method as an array of
+      # tuples, with a tuple for each parameter, e.g. `[[:req :a], [:opt
+      # :b]]`. Counting this array gives the arity.
+      arity = http_instance.class.instance_method(:get).parameters.count
```

Comment:
Do you know how fast this call is?  If this is measurably slow, I wonder if it would be better to memoize the parameter count.